### PR TITLE
Molten core - updates

### PR DESCRIPTION
**Pull request by cala**
- Fix melee behaviour for Ragnaros
- Rewrite Separation Anxiety for Garr and add same mechanics for Majordomo Executus

### CODE REVIEW

**Comment by cala** [comment]
Hi,

I'd like to have some feedback about that PR before merging it, namely from @xfurry but anyone is welcome to comment. =)

It fixes two things and both are now working as they should but the code needs review:
1. Ragnaros melee combat: previously he only checked if his victim was in melee reach before wiping the raid. Now he first check for his victim, then for another target in melee reach and finally he will cast Magma Blast if no one is found in melee. For this fix, I would like to know if the attack/target code is OK or if it could be optimised/improved
2. When pull far away from Garr, his adds gain a buff increasing their damage and making them immune to banish. In the previous code, this was handled by a hardcoded range check, each add casting the single buff onto himself, this buff has a 5 sec duration. This mechanics is now rewritten and move to Spell:Aura to use a persistent dummy aura cast by the boss. I scripted the dummy aura to have a 5 sec period. For each tick, the adds will check if they still are in the range of the aura from the boss. If not, they cast the 5 sec buff on themselves. The same code is used for Majordomo Executus which use the same mechanics for some of his adds. For this fix, I would like to know if the way the aura is scripted is OK or if this is some kind of hack.

Thanks for your review.

File: src/scriptdev2/scripts/eastern_kingdoms/molten_core/boss_ragnaros.cpp [hunk 1]
```diff
@@ -169,6 +167,58 @@ struct boss_ragnarosAI : public Scripted_NoMovementAI
...
+        }
+
+        // If a target is found in range
+        if (pTarget)
+        {
+            // Target is not current victim, force select and attack it
+            if (pTarget != m_creature->getVictim())
+            {
+                AttackStart(pTarget);
+                m_creature->SetTargetGuid(pTarget->GetObjectGuid());
```
Comment (killerwife) [inline src/scriptdev2/scripts/eastern_kingdoms/molten_core/boss_ragnaros.cpp]: Unit::Attack should take care of this. This happens in ScriptedAI::AttackStart. Isnt it redundant then?

File: src/scriptdev2/scripts/eastern_kingdoms/molten_core/boss_ragnaros.cpp [hunk 2]
```diff
@@ -169,6 +167,58 @@ struct boss_ragnarosAI : public Scripted_NoMovementAI
             m_uiEnterCombatTimer = 10000;
     }
 
+    // Custom threat management for targets in melee range
+    bool CanMeleeTargetInRange()
```
Comment (killerwife) [inline src/scriptdev2/scripts/eastern_kingdoms/molten_core/boss_ragnaros.cpp]: Cant this be substituted by SelectAttackingTarget - ATTACKING_TARGET_TOPAGGRO and SELECT_FLAG_IN_MELEE_RANGE?

File: src/game/SpellAuras.cpp [hunk 3]
```diff
@@ -1163,6 +1163,15 @@ void Aura::HandleAuraDummy(bool apply, bool Real)
                         if (Unit* caster = GetCaster())
                             caster->CastSpell(caster, 13138, TRIGGERED_OLD_TRIGGERED, nullptr, this);
                         return;
+					case 21094:								// Separation Anxiety (Majordomo Executus)
```
Comment (killerwife) [inline src/game/SpellAuras.cpp]: tabs

Comment (Konctantin) [inline src/game/SpellAuras.cpp]: need add .editorconfig

Comment (cala) [inline src/game/SpellAuras.cpp]: hmm... how did I miss that... Anyway, thanks, will fix.

File: src/scriptdev2/scripts/eastern_kingdoms/molten_core/boss_ragnaros.cpp (hunk 1 above)
Comment (cala) [inline src/scriptdev2/scripts/eastern_kingdoms/molten_core/boss_ragnaros.cpp]: That was my guess also... but for some unkown reason (at least to me), this does not happen, at least not all the time.
I will check again.

File: src/scriptdev2/scripts/eastern_kingdoms/molten_core/boss_ragnaros.cpp (hunk 2 above)
Comment (cala) [inline src/scriptdev2/scripts/eastern_kingdoms/molten_core/boss_ragnaros.cpp]: Do you mean the whole code to selec a target from line 172 to line 198 ?

**Comment by killerwife** [comment]
Yes. It seems to me that the code fulfills the same exact purpose.                                                                                                                                                                                                                                                                                                                                        Sent from my BlackBerry 10 smartphone on the Orange network.                                                                                                                                                                                                                From: calaSent: Monday, March 27, 2017 13:39To: cmangos/mangos-classicReply To: cmangos/mangos-classicCc: killerwife; CommentSubject: Re: [cmangos/mangos-classic] Molten core - updates (#274)@cala commented on this pull request.



In src/scriptdev2/scripts/eastern_kingdoms/molten_core/boss_ragnaros.cpp:
> @@ -169,6 +167,58 @@ struct boss_ragnarosAI : public Scripted_NoMovementAI
             m_uiEnterCombatTimer = 10000;
     }
 
+    // Custom threat management for targets in melee range
+    bool CanMeleeTargetInRange()

Do you mean the whole code to selec a target from line 172 to line 198 ?

—You are receiving this because you commented.Reply to this email directly, view it on GitHub, or mute the thread.


  
  




{"api_version":"1.0","publisher":{"api_key":"05dde50f1d1a384dd78767c55493e4bb","name":"GitHub"},"entity":{"external_key":"github/cmangos/mangos-classic","title":"cmangos/mangos-classic","subtitle":"GitHub repository","main_image_url":"https://cloud.githubusercontent.com/assets/143418/17495839/a5054eac-5d88-11e6-95fc-7290892c7bb5.png","avatar_image_url":"https://cloud.githubusercontent.com/assets/143418/15842166/7c72db34-2c0b-11e6-9aed-b52498112777.png","action":{"name":"Open in GitHub","url":"https://github.com/cmangos/mangos-classic"}},"updates":{"snippets":[{"icon":"PERSON","message":"@cala commented on #274"}],"action":{"name":"View Pull Request","url":"https://github.com/cmangos/mangos-classic/pull/274#discussion_r108148152"}}}

**Comment by killerwife** [comment]
There are a few conditions related to it, I have been wading through for some time now due to some targeting rules we dont follow from retail. There might be something shady going on that your code might unintentionally hide.                                                                                                                                                                                                                                                                                                                                        Sent from my BlackBerry 10 smartphone on the Orange network.                                                                                                                                                                                                                From: calaSent: Monday, March 27, 2017 13:35To: cmangos/mangos-classicReply To: cmangos/mangos-classicCc: killerwife; CommentSubject: Re: [cmangos/mangos-classic] Molten core - updates (#274)@cala commented on this pull request.



In src/scriptdev2/scripts/eastern_kingdoms/molten_core/boss_ragnaros.cpp:
> +                {
+                    Player* pPlayer = itr->getSource();
+                    if (pPlayer && pPlayer->isAlive() && !pPlayer->isGameMaster() && m_creature->CanReachWithMeleeAttack(pPlayer))
+                        pTarget = (Unit*)pPlayer;
+                }
+            }
+        }
+
+        // If a target is found in range
+        if (pTarget)
+        {
+            // Target is not current victim, force select and attack it
+            if (pTarget != m_creature->getVictim())
+            {
+                AttackStart(pTarget);
+                m_creature->SetTargetGuid(pTarget->GetObjectGuid());

That was my guess also... but for some unkown reason (at least to me), this does not happen, at least not all the time.
I will check again.

—You are receiving this because you commented.Reply to this email directly, view it on GitHub, or mute the thread.


  
  




{"api_version":"1.0","publisher":{"api_key":"05dde50f1d1a384dd78767c55493e4bb","name":"GitHub"},"entity":{"external_key":"github/cmangos/mangos-classic","title":"cmangos/mangos-classic","subtitle":"GitHub repository","main_image_url":"https://cloud.githubusercontent.com/assets/143418/17495839/a5054eac-5d88-11e6-95fc-7290892c7bb5.png","avatar_image_url":"https://cloud.githubusercontent.com/assets/143418/15842166/7c72db34-2c0b-11e6-9aed-b52498112777.png","action":{"name":"Open in GitHub","url":"https://github.com/cmangos/mangos-classic"}},"updates":{"snippets":[{"icon":"PERSON","message":"@cala commented on #274"}],"action":{"name":"View Pull Request","url":"https://github.com/cmangos/mangos-classic/pull/274#discussion_r108147454"}}}

File: src/scriptdev2/scripts/eastern_kingdoms/molten_core/boss_ragnaros.cpp (hunk 2 above)
Comment (cala) [inline src/scriptdev2/scripts/eastern_kingdoms/molten_core/boss_ragnaros.cpp]: Ok, I looked closer at Unit::SelectAttackingTarget and it should indeed do the trick. I'll test that soon and update the PR accordingly, thanks. 👍

**Comment by cala** [comment]
- Tabs are removed
- Ragnaros melee combat code is simplified: when something already exists in core, use it instead of rewriting it in scripts

Thanks.

When there is no more suggestions made about this PR, I'll merge it.

**Comment by cala** [comment]
Merged and rebased in https://github.com/cmangos/mangos-classic/commit/f345d21bfe652f5eea64f4203fd8bc87530d0447 and https://github.com/cmangos/mangos-classic/commit/14f445a5355e39df1035e77e93a38664c748c13a

**Pull request closed without merging**